drivers/led/led_gpio.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <drivers/gpio.h>
+#include <drivers/led.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(led_gpio, CONFIG_LED_GPIO_LOG_LEVEL);
+
+#define DT_DRV_COMPAT gpio_leds
+#define LED_DATA(node_id)                      \
+{                                              \
+	{                                      \
+		DT_GPIO_LABEL(node_id, gpios), \
+		DT_GPIO_PIN(node_id, gpios),   \
+		DT_GPIO_FLAGS(node_id, gpios)  \
+	},                                     \
+	DT_LABEL(node_id),                     \
+	DT_PROP(node_id, id)                   \
+},
+
+struct gpio_cfg {
+	const char *port;
+	gpio_pin_t pin;
+	gpio_dt_flags_t flags;
+};
+
+struct led_dts_cfg {
+	struct gpio_cfg gpio;
+	const char *label;
+	uint32_t id;
+};
+
+struct led_gpio_data {
+	struct device *gpio_dev;
+	const struct led_dts_cfg *cfg;
+	bool state;
+};
+
+struct led_gpio_cfg {
+	const struct led_dts_cfg *led;
+	uint32_t led_cnt;
+};
+
+static inline struct led_gpio_data *get_dev_data(struct device *dev)
+{
+	return dev->driver_data;
+}
+
+static inline const struct led_gpio_cfg *get_dev_config(struct device *dev)
+{
+	return dev->config_info;
+}
+
+static struct led_gpio_data *led_parameters_get(struct device *dev,
+						uint32_t led)
+{
+	const struct led_gpio_cfg *cfg = get_dev_config(dev);
+	struct led_gpio_data *data = get_dev_data(dev);
+
+	for (size_t i = 0; i < cfg->led_cnt; i++) {
+		if (data[i].cfg->id == led) {
+			return &data[i];
+		}
+	}
+
+	return NULL;
+}
+
+static int gpio_led_set_no_delay(struct led_gpio_data *data, bool state)
+{
+	int err;
+
+	data->state = state;
+
+	err = gpio_pin_set(data->gpio_dev, data->cfg->gpio.pin,
+			   state);
+	if (err) {
+		LOG_DBG("%s turn %s error %d",
+			data->cfg->label,
+			state ? "on" : "off",
+			err);
+	} else {
+		LOG_DBG("%s turn %s", data->cfg->label,
+			state ? "on" : "off");
+	}
+
+	return err;
+}
+
+static int gpio_led_set(struct led_gpio_data *data, bool state)
+{
+	return gpio_led_set_no_delay(data, state);
+}
+
+static int gpio_led_on(struct device *dev, uint32_t led)
+{
+	struct led_gpio_data *data = led_parameters_get(dev, led);
+
+	if (!data) {
+		return -EINVAL;
+	}
+
+	return gpio_led_set(data, true);
+}
+
+static int gpio_led_off(struct device *dev, uint32_t led)
+{
+	struct led_gpio_data *data = led_parameters_get(dev, led);
+
+	if (!data) {
+		return -EINVAL;
+	}
+
+	return gpio_led_set(data, false);
+}
+
+static const struct led_driver_api led_api = {
+	.on = gpio_led_on,
+	.off = gpio_led_off,
```

Comment:
I can define it but I would not remove this on/off callback I rather prefer to use it instead of `set_brightness` knowing that I use the GPIO LED driver where LED can be only on or off. It is less confusing for me

Reply:
Of course. I am just saying you could add the `set_brightness` callback.

Reply:
Maybe just add a callback that returns -ENOTSUP? The led.h API, at least as it is currently defined, requires set_brightness to be defined. So this is technically an API violation, even though I agree with @KAGA164 that it doesn't make sense for an LED connected to a GPIO to do anything with set_brightness.